modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionDemander.java

```diff
@@ -2006,47 +2006,43 @@ private void onChainFinished() {
-                    ", fullPartitions=" + fullParts +
-                    ", fullEntries=" + fullEntries +
-                    ", fullBytesRcvd=" + U.humanReadableByteCount(fullBytes) +
-                    ", topVer=" + topologyVersion() +
-                    ", progress=" + (routines - remainingRoutines) + "/" + routines + "]");
-            }
-            catch (Throwable t){
-                U.error(log, "Completed " + ((remainingRoutines == 0 ? "(final) " : "") +
-                    "rebalancing [grp=" + grp.cacheOrGroupName() +
-                    ", supplier=" + nodeId +
-                    ", topVer=" + topologyVersion() +
-                    ", progress=" + (routines - remainingRoutines) + "/" + routines + "]"), t);
-            }
+            // Node can left and by discovery we will not receive it.
+            Set<Integer> parts = nonNull(ctx.node(nodeId)) ? rebalancingParts.get(ctx.node(nodeId)) :
+                rebalancingParts.entrySet().stream().filter(e -> nodeId.equals(e.getKey().id()))
+                    .map(Map.Entry::getValue).findAny().orElse(Collections.emptySet());
+
+            Map<Boolean, Long> partCnts = parts.stream()
+                .collect(partitioningBy(historical::contains, counting()));
+
+            int fullParts = partCnts.getOrDefault(Boolean.FALSE, 0L).intValue();
+            int histParts = partCnts.getOrDefault(Boolean.TRUE, 0L).intValue();
+
+            long fullEntries = fullReceivedKeys.get(nodeId).sum();
+            long histEntries = histReceivedKeys.get(nodeId).sum();
+
+            long fullBytes = fullReceivedBytes.get(nodeId).sum();
+            long histBytes = histReceivedBytes.get(nodeId).sum();
+
+            long duration = System.currentTimeMillis() - startTime;
+            long durationSec = Math.max(1, TimeUnit.MILLISECONDS.toSeconds(duration));
+
+            U.log(log, "Completed " + (remainingRoutines == 0 ? "(final) " : "") +
+                "rebalancing [rebalanceId=" + rebalanceId +
+                ", grp=" + grp.cacheOrGroupName() +
+                ", supplier=" + nodeId +
+                ", partitions=" + (fullParts + histParts) +
+                ", entries=" + (fullEntries + histEntries) +
+                ", duration=" + U.humanReadableDuration(duration) +
+                ", bytesRcvd=" + U.humanReadableByteCount(fullBytes + histBytes) +
+                ", bandwidth=" + U.humanReadableByteCount((fullBytes + histBytes) / durationSec) + "/sec" +
+                ", histPartitions=" + histParts +
+                ", histEntries=" + histEntries +
+                ", histBytesRcvd=" + U.humanReadableByteCount(histBytes) +
+                ", fullPartitions=" + fullParts +
+                ", fullEntries=" + fullEntries +
+                ", fullBytesRcvd=" + U.humanReadableByteCount(fullBytes) +
+                ", topVer=" + topologyVersion() +
+                ", progress=" + (routines - remainingRoutines) + "/" + routines + "]");
```

Comment:
Are you 100% sure this code never throws any exception ? Let's keep the try - catch to avoid node shutdown on logging error. I suggest to do it for each complex logging block.